### ISSUE TITLE
MMS Image Quality Degradation 

### ISSUE THREAD

**Issue by Juush** [signalapp/Signal-Android#8590]
Let's begin with a checklist: Replace the empty checkboxes [ ] below with checked ones [x] accordingly. -->

- [x] I have searched open and closed issues for duplicates
- [X] I am submitting a bug report for existing functionality that does not work as intended
- [X] I have read https://github.com/signalapp/Signal-Android/wiki/Submitting-useful-bug-reports
- [X] This isn't a feature request or a discussion topic

----------------------------------------

### Bug description
I am trying to send MMS and they go through as normal but the quality degradation of the image is obscene. This is for screenshots and pictures taken through the app.

### Steps to reproduce
I have seen on other forum posts that this is very sporadic and hard to reproduce. 

For me, to reproduce it all I need to do is run the app and send a picture (either from my gallery) or from the in-app camera and send it off to someone. Clicking on the thumbnail shows how degregated it is, and is actually what the other person sees. This is for non-signal users. In the log, I have sent one to a signal user and a non-signal user.

**Actual result:** Resulting image is highly degregated and if text is on screen, illegible
**Expected result:** Sending MMS should suffer some quality loss but not completely lose the actual intended image.

This works just fine with other apps like Textra and the default android messenger, the incident is strictly in Signal. 

### Device info

**Device:** Pixel 3 XL 
**Android version:** 9
**Signal version:** 4.32.8

### Link to debug log

[Log Link](https://debuglogs.org/4c46df1a2faa2f4d16c5460f896cea7885fb0185adad3706a8a40174cef854f7)

**Comment by KonPaw** [comment]
So it looks like an issue similar to mine:
Every MMS I send have tragic quality while MMS I get are not bad. The same happened on my previous carrier.

I've checked previous issues, and one was fixed with: https://github.com/signalapp/Signal-Android/commit/51d61445915e0a607742a46c60332e015f9559b6
"4) Pull per-carrier MMS media constraints from the XML config
   files in the klinker library, instead of hardcoding it at 280kb."

I've checked klinker library (https://github.com/klinker41/android-smsmms) and as I can see, there are no xml files for my current and previous carrier (MCC 260, MNC 06, 03). My current carrier have limit of 600kB, it's a sad thing that I cannot change such a simple thing in the settings. As for argument that just send and email/signal as it's free - MMS-es here are also free, so if I want anyone to use Signal, it should work just as good as some slow, bad default messenger, or at least work better after some simple tweaking (as changing filesize in the settings), not work a lot worse... But i digress.  

I've did a small experiment, and downloaded some files (https://sample-videos.com/download-sample-jpg-image.php) and sent them to myself via mms.

Source JPG -> Received by MMS JPG
1MB 2192x2921 -> 15,23kB, 274x336
500kB 1792x1792 -> 29,98kB, 448x448
200kB 1036x1036 -> 12,86kB, 259x259 <- why make it worse?
100kB 689x689 -> 20,95kB, 345x345 <- why make it worse?
50KB 300x300 -> 15,66kB, 300x300 <- why make it worse?

After this i was sure, there was some werid 30kB limit. Than just to be sure, I've sent some computer generated file:

PNG
133kB 340x225 -> 133kB 340x225 <- that's the interesting one. It's some file generated by en.homm.fun (not an advert, some site with "astrologist proclaim" generator, looks fishy, checked by virus total and nothing found, but still enter with caution. I've just had this image on my phone).

I've tried with screenshots, as they are saved as PNG on my phone:
891kB 1440x2880 -> 22,94kB 180x360
1,68MB 1440x2880-> 22,44kB 180x360

Some goldhill.png test image:
792kB 720x576 -> 38.60kB 360x288

It seems to me that there is something really wrong with your image resizer, it goes for the overkill, even if it doesn't have to run in the first place. Unless it's some weird file, then it doesn't do this?

**Comment by Juush** [comment]
Hi KonPaw,

It definitely seems very similar. I can receive images just fine, it's just the sending. You obviously have a lot more technical knowledge than I do (I can pretty much understand what your tests were and the results... but that's the extent of it), but if you would like me to run some similar stuff to see if we get the same results, please do not hesitate to ask. I will try my best.

As an aside: 

> As for argument that just send and email/signal as it's free - MMS-es here are also free, so if I want anyone to use Signal, it should work just as good as some slow, bad default messenger, or at least work better after some simple tweaking (as changing filesize in the settings), not work a lot worse... But i digress.

I could not agree more, hence why surely this is a bug? I love the concept of Signal, and would happily encourage others to use it, especially with encryptions laws from my own government now encroaching on the day-to-day user, but I can hardly sway people to using it if I can hardly send a picture. I'm already an outlier by not having snapchat, shall I start asking my mates for their email so I can involve myself in these MMS discussions? 

In the mean time, I have gone back to Textra.

**Comment by KonPaw** [comment]
One thing that probably could help developers is what is your current carrier MCC and MNC, as it's probably the less known MCC/MNC carriers (that don't have optimal settings saved in Signal) that have this problem. By manually changing it in signal, they could try to reproduce it.

You can check it in APN settings on your phone.

Other than that, maybe check some pictures you sent via signal - before and after (if you select picture sent in your signal message, you can save it, and it will be the same that was delivered via MMS), to see the difference. Depending on you image viewer software you can then select it and get some info about it - most important is the resolution and size. In Google Photos, you can check it, when you click on "(i)".

Of course when someone will check this ticket, they will probably have more knowledge what they need to reproduce this ;)

**Comment by QuitPG** [comment]
Hi All,

Reddit links to the same problem:
https://www.reddit.com/r/signal/comments/atchzp/mms_images_too_smallcompressed_android/
https://www.reddit.com/r/signal/comments/8s5nxw/issues_with_mms_image_quality_and_receiving_text/
https://imgur.com/a/diZMsCt
https://www.reddit.com/r/signal/comments/agfxqd/extremely_poor_mms_quality_using_signal/

From a Oneplus 3 (Screenshot) (No XML match) - 270x480 36.59kb
From a Samsung S9 (Camera) (No XML match) - 189x252 21.00kb


I wouldn't mind an option of changing the limit (I think it varies a bit) depending on phone carrier, from my 2 second search I think 300kb is a standard International MMS limit but my carrier has a 500kb standard national limit but apparently can still send 2mb and receive 5mb (but not guaranteed). A hard 300kb International MMS limit might make sense but a flexible limit that can be set by the user (500kb) with correct resizing/compressing would be nice too.

Not the same phone and not using Signal but I have received images 1440x1080 227kb so it appears to just be when sending using Signal.





If I am understanding all this correctly then there are a couple of things wrong with the way this works.

-There are a lack of MCC MNC configs to be relying on them for settings - I found 1 for my country when there should be at least 2 others MINIMUM and potentially many more (1/3, 1/5, 1/10 isn't good enough). We could try to add a file for more of the carriers or maybe someone else already has and can be copied but I'm sure there are a lot of combinations.

-The default value with no XML file is way too aggressive and should be adjusted. With the reliance of XML files (lack of) then a smarter value should be set. Find a carrier that can't accept at least 100kb?? I think 300kb should be acceptable but I'm probably not aware of all the different carrier restrictions. Add a compatibility mode or opt in for higher quality if needed, use the values in the XML if they are correct (they are probably higher than 300kb, mine is 2mb like I thought was possible) if not use the higher default values.

-If this is true - "1024px limit for the longest side of an image" - It could probably be changed. I don't know if there is a problem with having an image too tall or wide other than compatibility with older phone that can't display high res images ---but my Nokia from like 10-15 years ago had a 2MP camera which means it should display fine so... I don't know, Where do you draw the line for compatibility.

TL:DR

MMS images may be getting compressed according to a matching of carrier specific XML files in the MMS library. If there isn't a file for your carrier, your images will be compressed to about 30kb.

We could:
-add more XML files for each carrier,
-edit the default (non XML match) to be higher, maybe add a compatibility mode just in case.
-allow the user to select their value.

This is limiting my ability to recommend Signal when I can't reliably send images.

**Comment by QuitPG** [comment]
There seems to be 3 versions of MMS - 1.1 (July 2004), 1.2 (May 2005) ,1.3 (September 2011).
MMS v1.2 may only support VGA resolution (640x480) at probably 100kb.

Page 17 and 18 and specs for the different classes for MMS v1.3.
https://www.openmobilealliance.org/release/MMS/V1_3-20110913-A/OMA-TS-MMS_CONF-V1_3-20110913-A.pdf

If targeting the correct class, MMS v1.3 spec is either 300kb or 600kb (up to the carrier) at up to 1600x1200.

An option to choose the MMS version would be good - v1.2 (compatibility), v1.3.

**Comment by QuitPG** [comment]
Is this the default XML file - https://github.com/klinker41/android-smsmms/blob/master/library/src/main/res/xml/mms_config.xml ?

It appears to be 640x480 like v1.2.
Looking at the resolutions of the files I have listed, it looks like Signal is trying to resize the image to under this value and even if it is under that size it still compresses it even though its not required.
KonPaw
> 50KB 300x300 -> 15,66kB, 300x300 <- why make it worse?
But the way it is resizing is by multiples of 2. 1080x1920 -> 270x480. Should 480 be the max since its 640x**480** or could/should it be 360x**640** (640x360)? -does the orientation of the image matter for MMS v1.2?
Also then, should Signal instead try to resize the exact pixel size instead of multiples to get the best quality out of a small size -especially if targeting MMS v1.2. I know it may be 'blurry' but I'm sure a 640x480 (360x640) 'blurry' resized image will look better than 189x252 exact resize. 

Signal appears to look for exact resolutions to get under 640x480 (not the other way around) by dividing by 2, 4, 8, 16
4032x3024 taken vertically = 3024x4032. 
3024x4032 -> /2 = 1512x2016
3024x4032 -> /4 = 756x1008 -- Would work with MMS v1.3
3024x4032 -> /6 = 504x672
3024x4032 -> /8 = 378x504 -- Could work with MMS v1.2 if 640x480 can be 480x640.
3024x4032 -> /10 = 302.4x403.2
3024x4032 -> /12 = 252x336
3024x4032 -> /14 = 216x288
3024x4032 -> /16 = 189x252

**Comment by patpilus** [comment]
I'm facing this issue for a long time.

For instance the screenshot from my phone with the original resolution of 1080x1575px (size 1.65 MB) is resized to 270x393px (60.37 KB). Resulting image is completely unreadable (text content). I don't get why the compression was set to be so high...

Device: Xiaomi Mi 6
Andorid version: 9.0.0 (custom ROM)
Signal version: 4.34.8

**Comment by QuitPG** [comment]
1080x1575 /2 = 540x787.5 (not under 640x480)
1080x1575 /4 = 270x393.75 (is under 640x480) - Signal chose this one.

Now to test my theory:
You appear to be from Poland, what is your mobile carrier/provider? Or more specifically the MCC and MNC?  ----If you are using a Poland sim, your MCC would be 260 just like KonPaw. There doesn't appear to be any XML files at all for any telco in Poland..... does that mean no one in Poland will be able to send an MMS above the Signal default of 640x480? 

Is my XML theory correct in any way? I can't seem to find the same files or the references to the carrier specific MMS settings with my searches in the Signal Github -only in the library linked earlier. I probably didn't search correctly but I'm guessing it must be in one of the database files. It just appears to be correct with my test of "1" but I can't really call that conclusive evidence yet.

I'm no real developer and won't be able to fix this myself but I want to get this sorted as soon as possible. I'm probably getting a lot wrong but I at least want to get this moving in the right direction.

**Comment by greyson-signal** [comment]
For anyone having this issue, the most useful information you can provide is your device info (phone, Android OS version) and what phone carrier you use.

**Comment by andreasmalling** [comment]
Phone: Nokia 7 plus (TA-1046) on Android 9 w. February security patch (build no.
00WW_3_39H)
Carrier: Call Me (MCC: 238, MNC: 20)

**Comment by billie80** [comment]
Device: samsung SM-G800H (kmini3gxx)
Android: 6.0.1 (G800HXXU1CPK3, MMB29M.G800HXXU1CPK3)
App: Signal 4.35.3
Carrier: HoT (MCC: 232, MNC: 07)

**Comment by billie80** [comment]
Just want to report back that with Signal 3.37.2 the image quality is much better now. I have a Dual Sim Phone and both Carriers allow 300 KB MMS. Previously an 3.1 MB (3264*1836) image was shrunk to 45 KB (408*230) for both SIM Cards. Now it is 250 KB (1024*576) for both. Thanks for taking care of this issue.

**Comment by alan-signal** [comment]
Thanks for the feedback @billie80

I believe that HoT MCC/MNC should allow 1600x1200.

So we can improve more, but it's great to hear you've seen some success so far!

**Comment by billie80** [comment]
Good to hear that there is still room for improvement. Btw. the other Carrier is Vodafone MCC 262  MNC 02.

Were do you get the resolution limits from? I only found out about the size limit.

**Comment by wingman1487** [comment]
Currently on the new 4.37.2 and still seeing large reduction in images and gifs (gif loose large amount of frames).
Phone: HTC U11
Carrier: Verizon
OS: Android 8.0.0

**Comment by alan-signal** [comment]
@wingman1487 You are talking about images you send I presume. Can I have more details.

Jpegs for example, what were the dimensions and file sizes being sent before for you and what are they now?

**Comment by wingman1487** [comment]
@alan-signal yes you are correct, images that I am sending.  Here are some quick ones, the GIF doesn't look like it is shrunk that much but there are tons of dropped frames that it doesn't look the same as the original.

|Photo|
Original:
9.0MP | 2250x4000 | 1.7MB
Sent:
2.1MP | 1094x1944 | 209kB

|GIF|
Original:
0.1MP | 500x281 | 424kB
Sent:
0.1 MP | 356x200 | 140kB

**Comment by greyson-signal** [comment]
@wingman1487 We don't touch gifs. If the recipient is seeing a gif of degraded quality, then maybe your carrier is compressing it.

**Comment by wingman1487** [comment]
@greyson-signal That would make sense, but interesting enough when using Signal to select a GIF from Giphy is when I first see the dropped frames.  When making the selection the GIFs play fine, however if I am sending to a non-Signal user once the GIF is selected and shown in the preview before sending you see the frame rate reduced.  This is not seen when sending to a Signal user, the preview prior to sending plays fine as long as Signal sees that the receiving party is also using Signal.
